### ISSUE TITLE
BuildStructure (or an alternative) needs to be public API

### ISSUE THREAD

**Issue by fommil** [sbt/sbt#2829]
I've been looking into upgrading `sbt-ensime` to sbt 1.0 and I've hit a few show stoppers with deprecated APIs.

Most notably the `BuildStructure` is now internal, but we use it extensively to inspect build information, see in https://github.com/ensime/ensime-sbt/blob/2.0/src/main/scala/EnsimePlugin.scala

`sbt-ensime` is primarily a `Command` that generates a `.ensime` file that contains explicit information about a build, such as source directories, class directories, jar deps, etc.

It is not possible (AFAIK) to write this as a `Task` because a Task would not be able to descend all the `projects` and query their state.

I suspect the intellij and scala-ide importer tools will face exactly the same problem.

**Comment by dwijnand** [comment]
I make use of `BuildStructure` (and related types) in sbt-project-graph too: 

https://github.com/dwijnand/sbt-project-graph/blob/bb5ccf8fa5fc6023e1a03cc5e8bc3ab355d9441b/src/main/scala/com/dwijnand/sbtprojectgraph/SbtProjectGraphPlugin.scala#L16

**Comment by eed3si9n** [comment]
BuildStructure probably should be superseded by proper commands and queries against the server's (JSON or Scala) API. We would want to make sure that import features continue to work for ENSIME and IDEs.

**Comment by xerial** [comment]
https://github.com/xerial/sbt-pack also depends on `BuildStructure` too.

**Comment by xerial** [comment]
@eed3si9n Is there any alternative way to recursively traverse project tasks?

**Comment by eed3si9n** [comment]
For the record, @xerial and I talked on Gitter channel.
For his use, I think we can just grab the list of all Projects first (`ReflectUtilities.allVals[Project](this).values.toList`), and then use `Extracted` to run tasks.

#3030

**Comment by xerial** [comment]
So far this works for both 0.13.x and 1.0.0-RC3
```
val extracted = Project.extract(state.value)
val structure = extracted.structure // Can get BuildStructure
```

**Comment by estk** [comment]
@fommil ^^^ I think @xerial 's solution should unblock https://github.com/ensime/ensime-sbt/issues/237 no?

**Comment by fommil** [comment]
we are using tasks now, so we might not have a problem. We do `Project.extract(s).structure` for sure

**Comment by dwijnand** [comment]
This is (also) tracked as #3296, btw.
